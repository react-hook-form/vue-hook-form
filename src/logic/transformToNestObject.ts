import set from '../utils/set';
import isKey from '../utils/isKey';
import { FieldValues } from '../types/form';

export default (data: FieldValues): any =>
  Object.entries(data).reduce(
    (previous: FieldValues, [key, value]): FieldValues => {
      if (!isKey(key)) {
        set(previous, key, value);
        return previous;
      }

      return { ...previous, [key]: value };
    },
    {},
  );
