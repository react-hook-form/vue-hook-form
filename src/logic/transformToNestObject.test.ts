import transformToNestObject from './transformToNestObject';

describe('transformToNestObject', () => {
  it('should combine all the array fields', () => {
    expect(
      transformToNestObject({
        'email[1]': 'asdasd@dsad.com',
        'email[2]': 'asdasd@.com',
        'firstName[1]': 'asdasd',
        'firstName[2]': 'asdasd',
        'lastName[1]': 'asdasd',
        'lastName[2]': 'asd',
        test: 'test',
      }),
    ).toMatchSnapshot();
  });

  it('should combine array object correctly', () => {
    expect(
      transformToNestObject({
        'name[0].firstName': 'testFirst',
        'name[0].lastName': 'testLast',
        'test[1].what': 'testLast',
        'test[1].task': 'testLast',
        'test[2].what': 'testLast',
        'test[2].what[1].test': 'testLast',
      }),
    ).toMatchSnapshot();
  });

  it('should combine object correctly', () => {
    expect(
      transformToNestObject({
        'name.firstName': 'testFirst',
        'name.lastName': 'testLast',
        'name.lastName.bill.luo': 'testLast',
      }),
    ).toMatchSnapshot();
  });

  it('should return default name value', () => {
    expect(
      transformToNestObject({
        name: 'testFirst',
      }),
    ).toMatchSnapshot();
  });

  it('should handle quoted values', () => {
    expect(
      transformToNestObject({
        'name["foobar"]': 'testFirst',
      }),
    ).toMatchSnapshot();

    expect(
      transformToNestObject({
        'name["b2ill"]': 'testFirst',
      }),
    ).toMatchSnapshot();
  });

  it('should combine with results', () => {
    expect(
      transformToNestObject({
        name: 'testFirst',
        name1: 'testFirst',
        name2: 'testFirst',
      }),
    ).toMatchSnapshot();
  });
});
