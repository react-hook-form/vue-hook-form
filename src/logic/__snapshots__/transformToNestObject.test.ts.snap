// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformToNestObject should combine all the array fields 1`] = `
Object {
  "email": Array [
    undefined,
    "asdasd@dsad.com",
    "asdasd@.com",
  ],
  "firstName": Array [
    undefined,
    "asdasd",
    "asdasd",
  ],
  "lastName": Array [
    undefined,
    "asdasd",
    "asd",
  ],
  "test": "test",
}
`;

exports[`transformToNestObject should combine array object correctly 1`] = `
Object {
  "name": Array [
    Object {
      "firstName": "testFirst",
      "lastName": "testLast",
    },
  ],
  "test": Array [
    undefined,
    Object {
      "task": "testLast",
      "what": "testLast",
    },
    Object {
      "what": Array [
        undefined,
        Object {
          "test": "testLast",
        },
      ],
    },
  ],
}
`;

exports[`transformToNestObject should combine object correctly 1`] = `
Object {
  "name": Object {
    "firstName": "testFirst",
    "lastName": Object {
      "bill": Object {
        "luo": "testLast",
      },
    },
  },
}
`;

exports[`transformToNestObject should combine with results 1`] = `
Object {
  "name": "testFirst",
  "name1": "testFirst",
  "name2": "testFirst",
}
`;

exports[`transformToNestObject should handle quoted values 1`] = `
Object {
  "name": Object {
    "foobar": "testFirst",
  },
}
`;

exports[`transformToNestObject should handle quoted values 2`] = `
Object {
  "name": Object {
    "b2ill": "testFirst",
  },
}
`;

exports[`transformToNestObject should return default name value 1`] = `
Object {
  "name": "testFirst",
}
`;
